tags/release_1_0b/snarf-7.0/snarf.1
.\" $Id$
.TH snarf 1 "17 Jun 2000"
.IX snarf
.SH NAME
snarf - Simple Non-interactive All-purpose Resource Fetcher
.SH SYNOPSIS
.B snarf
.RB [ -avqprzm ]
.B URL 
.RB [ outfile ]
.B ...
.PP

.SH DESCRIPTION
Retrieves data from a variety of protocols, namely 
.BR http ,
.BR ftp ,
and
.BR gopher .

.SH USAGE
.B snarf
is invoked with any number of URLs and outfiles. If an outfile is not
specified, snarf preserves the remote file name when saving.
.PP
For example,
.B snarf http://foo.bar.com/images/face.gif
will retrieve the file ``face.gif'' to the local system.  In the event
that there is no filename (the url ends in a slash), the data is
retrieved and stored in the file
.B index.html
for http URLs, 
.B ftpindex.txt
for ftp URLs, or
.B gopherindex.txt
for gopher URLs.

Using a dash, "-", as the outfile causes
.B snarf
to send its output to stdout rather than a file.
.PP
To log in to an ftp server or website that requires a username and
password, use the syntax http://username:password@site.com/. If you
omit the password, you will be prompted for it.
.PP
Snarf has a built-in option to download the latest version of itself;
simply run
.B snarf
.BR LATEST .


.SS OPTIONS
.TP
.I "\-a"
Causes snarf to use "active" ftp. By default, snarf uses passive ftp, 
and, if the server does not support it, falls back to active
ftp. Using the \-a option will avoid the initial passive attempt.
.TP
.I "\-r"
Resumes an interrupted ftp or http transfer by 
checking if there is a local file
with the same name as the remote file, and starting the transfer at the
end of the local file and continuing until finished. This option only
works with HTTP servers that understand HTTP/1.1 and ftp servers that
support the REST command. snarf uses this option automatically if the
outfile already exists.
.TP
.I "\-n"
Don't resume; ignore the outfile if it exists and re-transfer it in
its entirety.
.TP
.I "\-q"
Don't print progress bars.
.TP
.I "\-p"
Forces printing of progress bars. Snarf has a compile-time option for
whether progress bars print by default or not. The \-p option
overrides the \-q option. In addition, if progress bars are enabled by
default, snarf suppresses them when standard output is not a
terminal. Using
.I \-p
will override this behavior.
.TP
.I "\-v"
Prints all messages that come from the server to stderr.
.TP
.I "\-z"
Send a user-agent string similar to what Netscape Navigator 4.0 uses.
.TP
.I "\-m"
Send a user-agent string similar to what Microsoft Internet Explorer
uses.
.PP
Each option only affects the URL that immediately follows it. To have
an option affect all URLs that follow it, use an uppercase letter for
the option, e.g. \-Q instead of \-q.

.SH ENVIRONMENT
Snarf checks several environment variables when deciding what to use
for a proxy. It checks a service-specific variable first, then
.BR SNARF_PROXY ,
then
.BR PROXY .
.PP
The service-specific variables are
.BR HTTP_PROXY ,
.BR FTP_PROXY ,
and
.BR GOPHER_PROXY .

.PP
Snarf also checks the
.B SNARF_HTTP_USER_AGENT
environment variable and will use it when reporting its user-agent
string to an HTTP server. In the same spirit, it also uses the
.B SNARF_HTTP_REFERER
environment variable to spoof a Referer to the web server.

.SH BUGS
Bugs? What bugs? If you find 'em, report 'em. 
.SH AUTHOR
Copyright (C) 2000 Zachary Beane (xach@xach.com)
.\" Permission to use, copy, modify and distribute this software and its
.\" documentation for any purpose and without fee is hereby granted, provided
.\" that the above copyright notice appear in all copies and that both that
.\" copyright notice and this permission notice appear in supporting
.\" documentation.  This software is provided "as is" without express or
.\" implied warranty.


CHANGES
1.0d:
Sam Lantinga - Sun Nov 19 16:08:40 PST 2000
 * Added a UI message that shows up if a requested product is not installed
Sam Lantinga - Sun Nov 19 15:38:49 PST 2000
 * Added nice mirror control buttons for the download dialogs
Sam Lantinga - Sat Nov 18 16:56:40 PST 2000
 * Catch all updates for a product in an update list.
 * Added "Mirror:" keyword support in update lists.
Sam Lantinga - Sat Nov 18 15:58:55 PST 2000
 * Added arrow blinking at the end of a set of updates.
 * Added more detailed messages during update verification.
 * Added loading keyserver support (list in ~/.loki/loki_update/keyservers.txt)


snarf-7.0/snarf.c
/* -*- mode: C; c-basic-offset: 8; indent-tabs-mode: nil; tab-width: 8 -*- */
/* This program and all accompanying files are copyright 1998 Zachary
 * Beane <xach@xach.com>. They come with NO WARRANTY; see the file COPYING
 * for details.
 * 
 * This program is licensed to you under the terms of the GNU General
 * Public License. You should have recieved a copy of it with this
 * program; if you did not, you can view the terms at http://www.gnu.org/.
 *
 */

#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <pwd.h>
#include <string.h>
#include "options.h"
#include "llist.h"
#include "url.h"
#include "http.h"
#include "ftp.h"
#include "gopher.h"
#include "util.h"

#define LATEST_URL "http://www.xach.com/snarf/snarf-latest.tar.gz"

#ifdef PROGRESS_DEFAULT_OFF
#define PROGRESS_SETTING "off"
#else
#define PROGRESS_SETTING "on"
#endif

static void
usage(int verbose)
{
        if (!verbose) {
                fprintf(stderr, "Use `snarf --help' for help\n");
                exit(1);
        }

        printf("This is snarf, version %s\n", VERSION);
        printf("usage: snarf [OPTIONS] URL [OUTFILE] ...\n");
        printf("Options:\n"
               "    -a     Force active FTP (default is passive)\n"
               "    -v     Verbose; print anything the server sends\n" 
               "    -q     Don't print progress bars (compiled default is %s)\n"
               "    -p     Force printing of progress bars (overrides -q)\n"
               "    -r     Resume downloading a partially transferred file\n"
               "    -n     Ignore '-r' and transfer file in its entirety\n"
               "    -m     Spoof MSIE user-agent string\n"
               "    -z     Spoof Navigator user-agent string\n"
               "\n"
               "Lowercase option letters only affect the URLs that "
               "immediately follow them.\n"
               "If you give an option in caps, it will be the "
               "default option for all URLs\nthat follow it.\n"
               "\n"
               "If you specify the outfile as '-', the file will be "
               "printed to standard\noutput as it downloads.\n"
               "\n"
               "You can have as many URLs and outfiles as you want\n"
               "\n"
               "You can specify a username and password for ftp or "
               "http authentication. The\nformat is:\n"
               "\n"
               "    ftp://username:password@host/\n"
               "\n"
               "If you don't specify a password, you will be prompted "
                "for one.\n"
               "\n"
               "snarf checks the SNARF_PROXY, FTP_PROXY, GOPHER_PROXY, "
               "HTTP_PROXY, and PROXY\nenvironment variables.\n"
               "\n"
               "snarf is free software and has NO WARRANTY. See the file "
               "COPYING for details.\n", PROGRESS_SETTING);

        exit(1);
}
                

int
main(int argc, char *argv[])
{
        List *arglist		= NULL;
        UrlResource *rsrc	= NULL;
        Url *u			= NULL;
        int retval		= 0;
        int i;

        arglist = list_new();
        rsrc 	= url_resource_new();

        if(argc < 2) {
                fprintf(stderr, "snarf: not enough arguments\n");
                usage(0); /* usage() exits */
        }
        
        if (strcmp(argv[1], "--version") == 0) {
                printf("snarf %s\n", VERSION);
                exit(0);
        }

        if (strcmp(argv[1], "--help") == 0) {
                usage(1);
        }

        for(i = 1; --argc; i++) {
                if( strcmp(argv[i], "LATEST") == 0 ) {
                        u = url_new();
                        if( !url_init(u, LATEST_URL) ) {
                                report(NULL, ERR, "`%s' is not a valid URL");
                                continue;
                        }

                        rsrc->url = u;
                        continue;
                }

                /* options */
                if( argv[i][0] == '-' && argv[i][1] ) {
                        if( rsrc->url ) {
                                list_append(arglist, rsrc);
                                rsrc = url_resource_new();
                        }

                        rsrc->options = set_options(rsrc->options, argv[i]);
                        continue;
                }

                /* everything else */
                if( !rsrc->url ) {
                        u = url_new();
                        if( !url_init(u, argv[i]) ) {
                                report(NULL, ERR, "bad url `%s'", argv[i]);
                                continue;
                        }
                        rsrc->url = u;
                } else if( is_probably_an_url(argv[i]) ) {
                        list_append(arglist, rsrc);
                        rsrc = url_resource_new();

                        u = url_new();
                        if( !url_init(u, argv[i]) ) {
                                report(NULL, ERR, "bad url `%s'", argv[i]);
                                continue;
                        }
                        rsrc->url = u;
                } else {
                        if( rsrc->outfile )
                                report(NULL, WARN, "ignoring `%s' for outfile",
                                       argv[i]);
                        else
                                rsrc->outfile = strdup(argv[i]);
                }
        }

        if( rsrc->url ) {
                list_append(arglist, rsrc);
        }

        if (arglist->data == NULL) {
                fprintf(stderr, "snarf: not enough arguments\n");
                usage(0);
        }

        /* walk through the arglist and output and do that thing you
           do so well */

        while (arglist != NULL) {
                rsrc = arglist->data;
                arglist = arglist->next;
                u = rsrc->url;

                if( !rsrc->outfile && u->file )
                        rsrc->outfile = strdup(u->file);

                if( rsrc->outfile ) {
                        rsrc->outfile_offset = get_file_size(rsrc->outfile);
                        
                        if (rsrc->outfile_offset && 
                            !(rsrc->options & OPT_NORESUME)) {
                                fprintf(stderr, "setting resume on "
                                        "existing file `%s' at %ld bytes\n", 
                                        rsrc->outfile, rsrc->outfile_offset);
                                rsrc->options |= OPT_RESUME;
                        }
                }
                        

                if (u->username && !u->password) {
                        char *prompt = NULL;
                        
                        prompt = strconcat("Password for ",
                                           u->username, "@",
                                           u->host, ": ", NULL);
                        
                        u->password = strdup(getpass(prompt));
                        free(prompt);
                }

                i = transfer(rsrc);
                if (i == 0)
                        retval++;

                url_resource_destroy(rsrc);
        }
                
        return retval;
}
        
